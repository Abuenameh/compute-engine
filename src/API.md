# MathJSON API

## Parsing and Serializing

To transform Latex to MathJSON, use the `parse()` function.

To transform MathJSON to Latex, use the `serialize()` function.

```javascript
import { parse, serialize } from 'math-json';

const expr = parse('\\frac{\\pi}{2}');
console.log(expr);
// -> ["Divide", "Pi", 2]

const latex = serialize(expr);
console.log(latex);
// -> \frac{\pi}{2}
```

The behavior of parse and serialize can be customized by passing an optional
argument:

```javascript
import {  serialize } from 'math-json';

console.log(serialize(1/3, {
    precision: 3,
    decimalMarker: ","
}););
// -> 0,333
```

## Formating

A given mathematical expression can be represented in multiple equivalent ways
as a MathJSON expression. A **form** is used to specify a representation:

- **`'full'`**: only transformations applied are those necessary to make it
  valid JSON (for example making sure that `Infinity` and `NaN` are represented
  as strings)
- **`'flatten'`**: associative functions are combined, e.g. f(f(a, b), c) ->
  f(a, b, c)
- **`'sorted'`**: the arguments of commutative functions are sorted such that: -
  numbers are first, sorted numerically - complex numbers are next, sorted
  numerically by imaginary value - symbols are next, sorted lexicographically -
  `add` functions are next - `multiply` functions are next - `power` functions
  are next, sorted by their first argument, then by their second argument -
  other functions follow, sorted lexicographically
- **`'stripped-metadata'`**: any metadata associated with elements of the
  expression is removed.
- **`'object-literal'`**: each term of an expression is expressed as an object
  literal: no shorthand representation is used.
- **`'canonical-add'`**: `addition of 0 is simplified, associativity rules are
  applied, unnecessary groups are moved, single argument 'add' are simplified
- **`'canonical-divide'`**: `divide` is replaced with `multiply` and `power',
  division by 1 is simplified,
- **`'canonical-exp'`**: `exp` is replaced with `power`
- **`'canonical-multiply'`**: multiplication by 1 or -1 is simplified
- **`'canonical-power'`**: `power` with a first or second argument of 1 is
  simplified
- **`'canonical-negate'`**: real or complex number is replaced by the negative
  of that number. Negation of negation is simplified.
- **`'canonical-number'`**: complex numbers with no imaginary compnents are
  simplified
- **`'canonical-root'`**: `root` is replaced with `power`
- **`'canonical-subtract'`**: `subtract` is replaced with `add` and `negate`
- **`'canonical'`**: the following transformations are performed, in this order:
  - 'canonical-number', // -> simplify number
  - 'canonical-exp', // -> power
  - 'canonical-root', // -> power, divide
  - 'canonical-subtract', // -> add, negate, multiply,
  - 'canonical-divide', // -> multiply, power
  - 'canonical-power', // simplify power
  - 'canonical-multiply', // -> multiply, power
  - 'canonical-negate', // simplify negate
  - 'canonical-add', // simplify add
  - 'flatten', // simplify associative, idempotent and groups
  - 'sorted',
  - 'full',

To transform an expression using the rules for a particular form, use the
`format()` function.

```js
import { format } from 'math-json';

console.log(format(["Add", 2, "x", 3], ['canonical']);
// -> ["Add", 2, 3, "x"]
```

## Advanced Usage

To improve performance, particularly when calling `parse()`/`serialize()`
repeatedly, use an instance of the `LatexSyntax` class. When the instance is
constructed, the dictionaries defining the syntax are compiled, and subsequent
invocations of the `parse()` and `serialize()` methods can skip that step.

```javascript
import { LatexSyntax } from 'math-json';
const latexSyntax = new LatexSyntax();
const expr = latexSyntax.parse('\\frac{\\pi}{2}');
console.log(expr);
const latex = latexSyntax.serialize(expr);
console.log(latex);
```

To customize the syntax, provide options to the constructor of `LatexSyntax`.

For example, the configuration below will result in parsing a Latex string as a
sequence of Latex tokens.

```js
const rawLatex = new LatexSyntax({
  parseArgumentsOfUnknownLatexCommands: false,
  promoteUnknownSymbols: /./,
  invisibleOperator: '',
  invisiblePlusOperator: '',
  dictionary: [],
  skipSpace: false,
});
const expr = rawLatex.parse('\\frac{\\pi}{2}');
console.log(expr);
// -> ["Latex", "\frac", "<{>", "\pi", "<}>", "<{>",  2, "<}>"]
```

Similarly, create an instance of `ComputeEngine` for calls to `format()` and
`evaluate()`.
