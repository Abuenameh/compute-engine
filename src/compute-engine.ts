// Compute Engine
export {
  ComputeEngine,
  format,
  evaluate,
} from './compute-engine/compute-engine';
