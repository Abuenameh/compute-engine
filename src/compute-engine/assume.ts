import { AssumeResult, BoxedExpression, IComputeEngine } from './public';

import { findUnivariateRoots } from './solve';

/**
 * Add an assumption, in the form of a predicate, for example:
 *
 * - `x = 5`
 * - `x ∈ ℕ`
 * - `x > 3`
 * - `x + y = 5`
 *
 * Some assumptions are handled separately, specifically, those that can
 * be represented as a symbol definition (equality to an expression,
 * membership to Integer, RealNumber, etc..., >0, <=0, etc...). The result
 * of these are stored directly in the current scope's symbols dictionary
 * (and an entry for the symbol is created if necessary).
 *
 * New assumptions can 'refine' previous assumptions, that is they are valid
 * if they don't contradict previous assumptions. To set new assumptions
 * that contradict previous ones, you must first `forget` about any symbols
 * in the new assumption.
 *
 * Predicates that involve multiple symbols are simplified (for example
 * `x + y = 5` becomes `x + y - 5 = 0`, then stored in the `assumptions` of the
 * current context).
 *
 */

export function assume(proposition: BoxedExpression): AssumeResult {
  if (proposition.head === 'Element') return assumeElement(proposition);
  if (proposition.head === 'Equal') return assumeEquality(proposition);
  if (isInequality(proposition)) return assumeInequality(proposition);

  return 'not-a-predicate';
}

function assumeEquality(proposition: BoxedExpression): AssumeResult {
  console.assert(proposition.head === 'Equal');
  // Four cases:
  // 1/ proposition contains no free variables
  //    e.g. `2 + 1 = 3`, `\pi + 1 = \pi`
  //    => evaluate and return
  // 2/ lhs is a single free variable and `rhs` does not contain `lhs`
  //    e.g. `x = 2`, `x = 2\pi`
  //    => if `lhs` has a definition, set its value to `rhs`, otherwise
  //          define a new symbol with a value of `rhs`
  // 3/ proposition contains a single free variable
  //    => solve for the free variable, create new def or set value of the
  //      free variable with the root(s) as value
  // 4/ proposition contains multiple free variables
  //    => add (lhs - rhs = 0) to assumptions DB

  // Case 1
  const freeVars = proposition.freeVars;
  if (freeVars.length === 0) {
    const val = proposition.evaluate();
    if (val.symbol === 'True') return 'tautology';
    if (val.symbol === 'False') return 'contradiction';
    console.log(proposition.canonical.evaluate());
    return 'not-a-predicate';
  }

  const ce = proposition.engine;

  // Case 2
  const lhs = proposition.op1.symbol;
  if (lhs && !hasValue(ce, lhs) && !proposition.op2.has(lhs)) {
    const val = proposition.op2.evaluate();
    if (!val.isValid) return 'not-a-predicate';
    const def = ce.lookupSymbol(lhs);
    if (!def) {
      ce.defineSymbol(lhs, { value: val });
      return 'ok';
    }
    if (def.domain && !val.domain.isCompatible(def.domain))
      return 'contradiction';
    def.value = val;
    return 'ok';
  }

  // Case 3
  if (freeVars.length === 1) {
    const lhs = freeVars[0];
    const sols = findUnivariateRoots(proposition, lhs);
    if (sols.length === 0) {
      ce.assumptions.set(
        ce.fn('Equal', [
          ce
            .add([proposition.op1.canonical, ce.neg(proposition.op2.canonical)])
            .simplify(),
          0,
        ]),
        true
      );
    }

    const val = sols.length === 1 ? sols[0] : ce.fn('List', sols);
    const def = ce.lookupSymbol(lhs);
    if (!def) {
      ce.defineSymbol(lhs, { value: val });
      return 'ok';
    }
    if (def.domain && !sols.every((sol) => val.domain.isCompatible(sol.domain)))
      return 'contradiction';
    def.value = val;
    return 'ok';
  }

  ce.assumptions.set(proposition, true);
  return 'ok';
}

function assumeInequality(proposition: BoxedExpression): AssumeResult {
  //
  // 1/ lhs is a single **undefined** free var e.g. "x < 0"
  //    => define a new var, if the domain can be inferred set it, otherwise
  // RealNumber and add to assumptions (e.g. x < 5)
  // 2/ (lhs - rhs) is an expression with no free vars
  //  e.g. "\pi < 5"
  //  => evaluate
  // 3/ (lhs - rhs) is an expression with a single **undefined** free var
  //    e.g. "x + 1 < \pi"
  //    => add def as RealNumber, add to assumptions
  // 4/ (lhs - rhs) is an expression with multiple free vars
  //    e.g. x + y < 0
  //    => add to assumptions

  const ce = proposition.engine;
  // Case 1
  if (proposition.op1!.symbol && !hasDef(ce, proposition.op1!.symbol)) {
    if (proposition.op2.evaluate().isZero) {
      if (proposition.head === 'Less') {
        ce.defineSymbol(proposition.op1.symbol, {
          domain: ce.domain('NegativeNumber'),
        });
      } else if (proposition.head === 'LessEqual') {
        ce.defineSymbol(proposition.op1.symbol, {
          domain: ce.domain('NonPositiveNumber'),
        });
      } else if (proposition.head === 'Greater') {
        ce.defineSymbol(proposition.op1.symbol, {
          domain: ce.domain('PositiveNumber'),
        });
      } else if (proposition.head === 'GreaterEqual') {
        ce.defineSymbol(proposition.op1.symbol, {
          domain: ce.domain('NonNegativeNumber'),
        });
      }
    } else {
      ce.defineSymbol(proposition.op1.symbol, {
        domain: ce.domain('ExtendedRealNumber'),
      });
      ce.assumptions.set(proposition, true);
    }
    return 'ok';
  }
  // @todo: handle if proposition.op1 *has* a def (and no value)

  // Normalize to Less, LessEqual
  let op = '';
  let lhs: BoxedExpression;
  let rhs: BoxedExpression;
  if (proposition.head === 'Less') {
    lhs = proposition.op1;
    rhs = proposition.op2;
    op = '<';
  } else if (proposition.head === 'LessEqual') {
    lhs = proposition.op1;
    rhs = proposition.op2;
    op = '<=';
  } else if (proposition.head === 'Greater') {
    lhs = proposition.op2;
    rhs = proposition.op1;
    op = '<';
  } else if (proposition.head === 'GreaterEqual') {
    lhs = proposition.op2;
    rhs = proposition.op1;
    op = '<=';
  }
  if (!op) return 'internal-error';
  const p = ce.add([lhs!.canonical, ce.neg(rhs!.canonical)]).simplify();

  // Case 2
  const result = ce.box([op === '<' ? 'Less' : 'LessEqual', p, 0]).evaluate();

  if (result.symbol === 'True') return 'tautology';
  if (result.symbol === 'False') return 'contradiction';

  const freeVars = result.freeVars;
  if (freeVars.length === 0) return 'not-a-predicate';

  // Case 3
  if (freeVars.length === 1) {
    if (!ce.lookupSymbol(freeVars[0]))
      ce.defineSymbol(freeVars[0], { domain: 'ExtendedRealNumber' });
  }

  // Case 3, 4
  console.assert(result.head === 'Less' || result.head === 'LessEqual');
  ce.assumptions.set(result, true);
  return 'ok';
}

function assumeElement(proposition: BoxedExpression): AssumeResult {
  console.assert(proposition.head === 'Element');

  // Four cases:
  // 1/ lhs is a single free variable with no definition
  //    e.g. `x \in \R`
  //    => define a new var with the specified domain
  // 2/ lhs is a symbol with a definition
  //    => update domain, if compatible
  // 3/ lhs is an expression with some free variables with no definition
  //    => add to assumptions DB
  // 4/ otherwise  (expression)
  //    e.g. `x+2 \in \R`
  //    => evaluate and return result (contradiction or tautology)

  const ce = proposition.engine;
  const undefs = undefinedIdentifiers(proposition.op1);
  // Case 1
  if (undefs.length === 1) {
    const dom = ce.domain(proposition.op2.evaluate().json);
    if (!dom.isValid) return 'not-a-predicate';
    if (dom.isCompatible('Function'))
      ce.defineFunction(undefs[0], { signature: { domain: 'Function' } });
    else ce.defineSymbol(undefs[0], { domain: dom });
    return 'ok';
  }

  // Case 2
  if (proposition.op1.symbol && hasDef(ce, proposition.op1.symbol)) {
    const dom = ce.domain(proposition.op2.evaluate().json);
    if (!dom.isValid) return 'not-a-predicate';
    const def = ce.lookupSymbol(proposition.op1.symbol);
    if (def) {
      if (def.domain && !dom.isCompatible(def.domain)) return 'contradiction';
      def.domain = dom;
      return 'ok';
    }
    const fdef = ce.lookupFunction(proposition.op1.symbol);
    if (fdef?.signature?.domain) {
      if (!dom.isCompatible(fdef.signature.domain)) return 'contradiction';
      if (dom.isCompatible(fdef.signature.domain, 'bivariant'))
        return 'tautology';
      return 'not-a-predicate';
    }
    return 'ok';
  }

  // Case 3
  if (undefs.length > 0) {
    ce.assumptions.set(proposition, true);
    return 'ok';
  }

  // Case 4
  const val = proposition.evaluate();
  if (val.symbol === 'True') return 'tautology';
  if (val.symbol === 'False') return 'contradiction';
  return 'not-a-predicate';
}

function hasDef(ce: IComputeEngine, s: string): boolean {
  return (ce.lookupSymbol(s) ?? ce.lookupFunction(s)) !== undefined;
}

function undefinedIdentifiers(expr: BoxedExpression): string[] {
  return expr.symbols.filter((x) => !hasDef(expr.engine, x));
}

function hasValue(ce: IComputeEngine, s: string): boolean {
  if (ce.lookupFunction(s)) return false;
  return ce.lookupSymbol(s)?.value !== undefined;
}

function isInequality(expr: BoxedExpression): boolean {
  const h = expr.head;
  if (typeof h !== 'string') return false;
  return ['Less', 'Greater', 'LessEqual', 'GreaterEqual'].includes(h);
}
