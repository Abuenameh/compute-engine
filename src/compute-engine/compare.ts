import { ComputeEngine, Expression } from '../public';
import { order } from './order';

/**
 * Structural comparison between two expressions
 */
export function compare(
  _engine: ComputeEngine,
  lhs: Expression,
  rhs: Expression
): -1 | 0 | 1 | undefined {
  // @todo
  // Special case if both numbers
  // Special case if both domains
  // Special case if both sets
  // What to do with Boolean? Not orderable, but Expressions are...
  // If both Expressions (general case)
  const result = order(lhs, rhs);

  return result < 0 ? -1 : result > 0 ? +1 : 0;
}
