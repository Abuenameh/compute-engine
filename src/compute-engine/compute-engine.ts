import {
  DictionaryCategory,
  ErrorSignal,
  Expression,
  WarningSignal,
} from '../public';

import {
  CollectionDefinition,
  ComputeEngine as ComputeEngineInterface,
  Definition,
  Dictionary,
  Domain,
  Form,
  FunctionDefinition,
  Numeric,
  NumericFormat,
  RuleSet,
  RuntimeScope,
  Scope,
  SetDefinition,
  Simplification,
  SymbolDefinition,
} from './public';

import { InternalComputeEngine } from './internal-compute-engine';
import { equalExpr } from '../common/utils';
import { ExpressionMap } from './expression-map';

export class ComputeEngine<T extends number = number>
  implements ComputeEngineInterface
{
  readonly internal: InternalComputeEngine;

  static getDictionaries(
    categories: DictionaryCategory[] | 'all' = 'all'
  ): Readonly<Dictionary<Numeric>>[] {
    return InternalComputeEngine.getDictionaries(categories);
  }

  constructor(options?: { dictionaries?: Readonly<Dictionary<Numeric>>[] }) {
    this.internal = new InternalComputeEngine(options);
  }

  get precision(): number {
    return this.internal.precision;
  }
  set precision(p: number | 'machine') {
    this.internal.precision = p;
  }

  get numericFormat(): NumericFormat {
    return this.internal.numericFormat;
  }
  set numericFormat(f: NumericFormat) {
    this.internal.numericFormat = f;
  }

  get context(): RuntimeScope<Numeric> {
    return this.internal.context;
  }

  pushScope(
    dictionary: Readonly<Dictionary<Numeric>>,
    scope?: Partial<Scope>
  ): void {
    this.internal.pushScope(dictionary, scope);
  }

  popScope(): void {
    this.internal.popScope();
  }

  get assumptions(): ExpressionMap<T, boolean> {
    return this.internal.assumptions;
  }

  shouldContinueExecution(): boolean {
    return this.internal.shouldContinueExecution();
  }

  checkContinueExecution(): void {
    this.internal.checkContinueExecution();
  }

  signal(sig: ErrorSignal | WarningSignal): void {
    this.internal.signal(sig);
  }

  getFunctionDefinition(name: string): FunctionDefinition | null {
    return this.internal.getFunctionDefinition(name);
  }
  getSymbolDefinition(name: string): SymbolDefinition<Numeric> | null {
    return this.internal.getSymbolDefinition(name);
  }
  getSetDefinition(name: string): SetDefinition<Numeric> | null {
    return this.internal.getSetDefinition(name);
  }
  getCollectionDefinition(name: string): CollectionDefinition<Numeric> | null {
    return this.internal.getCollectionDefinition(name);
  }
  getDefinition(name: string): Definition<Numeric> | null {
    return this.internal.getDefinition(name);
  }

  getRules(topic: string | string[]): RuleSet {
    return this.internal.getRules(topic);
  }

  format(expr: Expression | null, forms?: Form | Form[]): Expression | null {
    return this.internal.format(expr, forms);
  }

  /**
   * Return the canonical form of an expression.
   */
  canonical(expr: Expression | null): Expression | null {
    return this.internal.format(expr, ['canonical']);
  }

  /**
   * Return a numerical approximation of an expression.
   */
  N(expr: Expression): Expression | null {
    return this.internal.N(expr);
  }

  /**
   * Attempt to simplify an expression, that is rewrite it in a simpler form,
   * making use of the available assumptions.
   *
   * The simplification steps will proceed multiple times until either:
   * 1/ the expression stop changing
   * 2/ the number of iteration exceeds `iterationLimit`
   * 3/ the time to compute exceeds `timeLimit`, expressed in seconds
   *
   * If no `timeLimit` or `iterationLimit` are provided, the values
   * from the current ComputeEngine context are used. By default those
   * values are an infinite amount of iterations and a 2s time limit.
   *
   */
  simplify(
    expr: Expression,
    options?: {
      timeLimit?: number;
      iterationLimit?: number;
      simplifications?: Simplification[];
    }
  ): Expression | null {
    const simplifications: Simplification[] = options?.simplifications ?? [
      'all',
    ];
    const timeLimit = options?.timeLimit ?? this.internal.timeLimit ?? 2.0;
    if (timeLimit && isFinite(timeLimit)) {
      this.internal.deadline = Date.now() + timeLimit * 1000;
    }

    const iterationLimit =
      options?.iterationLimit ?? this.internal.iterationLimit ?? 1024;
    let iterationCount = 0;
    let result: Expression | null = this.canonical(expr);
    let prevResult: Expression | null = result;
    while (iterationCount < iterationLimit && this.shouldContinueExecution()) {
      result = this.internal.simplify(result!, { simplifications });
      if (result === null) return prevResult;
      if (equalExpr(prevResult, result)) return this.canonical(result);
      prevResult = result;
      iterationCount += 1;
    }
    if (result === null) return null;
    return this.canonical(result);
  }

  /**
   * Return a simplified and numerically approximation of an expression
   * in canonical form.
   *
   * The simplification steps will proceed multiple times until either:
   * 1/ the expression stop changing
   * 2/ the number of iteration exceeds `iterationLimit`
   * 3/ the time to compute exceeds `timeLimit`, expressed in seconds
   *
   * If no `timeLimit` or `iterationLimit` are provided, the values
   * from the current ComputeEngine context are used. By default those
   * values are an infinite amount of iterations and a 2s time limit.
   */
  async evaluate(
    expr: Expression,
    options?: { timeLimit?: number; iterationLimit?: number }
  ): Promise<Expression | null> {
    const val = this.internal.evaluate(expr, options);
    if (val === null) return null;
    return this.canonical(await val);
  }

  parse(s: string): Expression {
    return this.internal.parse(s);
  }
  serialize(x: Expression): string {
    return this.serialize(x);
  }

  domain(expr: Expression): Expression | null {
    return this.internal.domain(expr);
  }

  /** Query the assumption database
   */
  is(symbol: Expression, domain: Domain): boolean | undefined;
  is(predicate: Expression): boolean | undefined;
  is(arg1: Expression, arg2?: Domain): boolean | undefined {
    return this.internal.is(arg1, arg2);
  }

  ask(pattern: Expression): { [symbol: string]: Expression }[] {
    return this.internal.ask(pattern);
  }

  assume(
    symbol: Expression,
    domain: Domain
  ): 'contradiction' | 'tautology' | 'ok';
  assume(predicate: Expression): 'contradiction' | 'tautology' | 'ok';
  assume(
    arg1: Expression,
    arg2?: Domain
  ): 'contradiction' | 'tautology' | 'ok' {
    return this.internal.assume(arg1, arg2);
  }

  replace(rules: RuleSet<T>, expr: Expression<T>): Expression<T> {
    return this.internal.replace(rules, expr);
  }

  getVars(expr: Expression): Set<string> {
    return this.internal.getVars(expr);
  }

  isZero(x: Expression<T>): boolean | undefined {
    return this.internal.isZero(x);
  }
  isNotZero(x: Expression<T>): boolean | undefined {
    return this.internal.isNotZero(x);
  }
  isNumeric(x: Expression<T>): boolean | undefined {
    return this.internal.isNumeric(x);
  }
  isInfinity(x: Expression<T>): boolean | undefined {
    return this.internal.isInfinity(x);
  }
  // Not +- Infinity, not NaN
  isFinite(x: Expression<T>): boolean | undefined {
    return this.internal.isFinite(x);
  }
  // x >= 0
  isNonNegative(x: Expression<T>): boolean | undefined {
    return this.internal.isNonNegative(x);
  }
  // x > 0
  isPositive(x: Expression<T>): boolean | undefined {
    return this.internal.isPositive(x);
  }
  // x < 0
  isNegative(x: Expression<T>): boolean | undefined {
    return this.internal.isNegative(x);
  }
  // x <= 0
  isNonPositive(x: Expression<T>): boolean | undefined {
    return this.internal.isNonPositive(x);
  }
  isInteger(x: Expression<T>): boolean | undefined {
    return this.internal.isInteger(x);
  }
  isRational(x: Expression<T>): boolean | undefined {
    return this.internal.isRational(x);
  }
  isAlgebraic(x: Expression<T>): boolean | undefined {
    return this.internal.isAlgebraic(x);
  }
  isReal(x: Expression<T>): boolean | undefined {
    return this.internal.isReal(x);
  }
  // Real or +-Infinity
  isExtendedReal(x: Expression<T>): boolean | undefined {
    return this.internal.isExtendedReal(x);
  }
  isComplex(x: Expression<T>): boolean | undefined {
    return this.internal.isComplex(x);
  }
  isOne(x: Expression<T>): boolean | undefined {
    return this.internal.isOne(x);
  }
  isNegativeOne(x: Expression<T>): boolean | undefined {
    return this.internal.isNegativeOne(x);
  }
  isElement(x: Expression<T>, set: Expression<T>): boolean | undefined {
    return this.internal.isElement(x, set);
  }
  isSubsetOf(lhs: Domain | null, rhs: Domain | null): boolean {
    return this.internal.isSubsetOf(lhs, rhs);
  }

  isEqual(lhs: Expression, rhs: Expression): boolean | undefined {
    return this.internal.isEqual(lhs, rhs);
  }
  isLess(lhs: Expression, rhs: Expression): boolean | undefined {
    return this.internal.isLess(lhs, rhs);
  }
  isLessEqual(lhs: Expression, rhs: Expression): boolean | undefined {
    return this.internal.isLessEqual(lhs, rhs);
  }
  isGreater(lhs: Expression, rhs: Expression): boolean | undefined {
    return this.internal.isGreater(lhs, rhs);
  }
  isGreaterEqual(lhs: Expression, rhs: Expression): boolean | undefined {
    return this.internal.isGreaterEqual(lhs, rhs);
  }
}

let gComputeEngine: ComputeEngine | null = null;

export function format(
  expr: Expression,
  forms: Form | Form[]
): Expression | null {
  if (gComputeEngine === null) gComputeEngine = new ComputeEngine();
  return gComputeEngine.format(expr, forms);
}

export function evaluate(expr: Expression): Promise<Expression | null> {
  if (gComputeEngine === null) gComputeEngine = new ComputeEngine();
  return gComputeEngine.evaluate(expr);
}
