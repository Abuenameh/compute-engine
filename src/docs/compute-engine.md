---
title: MathJSON Core
permalink: /guides/math-json-compute-engine-api/
layout: single
date: Last Modified
sidebar:
  - nav: 'mathjson'
---

# Compute Engine API

## Formating

A given mathematical expression can be represented in multiple equivalent ways
as a MathJSON expression. A **form** is used to specify a representation:

- **`'full'`**: only transformations applied are those necessary to make it
  valid JSON (for example making sure that `Infinity` and `NaN` are represented
  as strings)
- **`'flatten'`**: associative functions are combined, e.g. f(f(a, b), c) ->
  f(a, b, c)
- **`'sorted'`**: the arguments of commutative functions are sorted such that: -
  numbers are first, sorted numerically - complex numbers are next, sorted
  numerically by imaginary value - symbols are next, sorted lexicographically -
  `add` functions are next - `multiply` functions are next - `power` functions
  are next, sorted by their first argument, then by their second argument -
  other functions follow, sorted lexicographically
- **`'stripped-metadata'`**: any metadata associated with elements of the
  expression is removed.
- **`'object-literal'`**: each term of an expression is expressed as an object
  literal: no shorthand representation is used.
- **`'canonical-add'`**: `addition of 0 is simplified, associativity rules are
  applied, unnecessary groups are moved, single argument 'add' are simplified
- **`'canonical-divide'`**: `divide` is replaced with `multiply` and `power',
  division by 1 is simplified,
- **`'canonical-exp'`**: `exp` is replaced with `power`
- **`'canonical-multiply'`**: multiplication by 1 or -1 is simplified
- **`'canonical-power'`**: `power` with a first or second argument of 1 is
  simplified
- **`'canonical-negate'`**: real or complex number is replaced by the negative
  of that number. Negation of negation is simplified.
- **`'canonical-number'`**: complex numbers with no imaginary compnents are
  simplified
- **`'canonical-root'`**: `root` is replaced with `power`
- **`'canonical-subtract'`**: `subtract` is replaced with `add` and `negate`
- **`'canonical'`**: the following transformations are performed, in this order:
  - `'canonical-number'`, -> simplify number
  - `'canonical-exp'`, -> power
  - `'canonical-root'`, -> power, divide
  - `'canonical-subtract'`, -> add, negate, multiply,
  - `'canonical-divide'`, -> multiply, power
  - `'canonical-power'`, -> simplify power
  - `'canonical-multiply'`, -> multiply, power
  - `'canonical-negate'`, -> simplify negate
  - `'canonical-add'`, -> simplify add
  - `'flatten'`, simplify associative, idempotent, involution and groups
  - `'sorted'`,
  - `'full'`,

To transform an expression using the rules for a particular form, use the
`format()` function.

```js
import { format } from 'math-json';

console.log(format(["Add", 2, "x", 3], 'canonical');
// -> ["Add", 2, 3, "x"]
```

## Evaluating

```js
import { evaluate } from 'math-json';

console.log(evaluate(["Add", 2, 3]);
// -> 5
console.log(evaluate(["Add", 2, "x", 3]);
// -> ["Add", 5, x]
```

## Advanced Usage

To improve performance, particularly when calling `format()`/`evaluate()`
repeatedly, use an instance of the `ComputeEngine` class. When the instance is
constructed, the dictionaries defining the symbols are compiled, and subsequent
invocations of the `format()` and `evaluate()` methods can skip that step.

Using a compute engine instance, it is possible to customize which symbol
dictionaries are used.

```js
const engine = new ComputeEngine(ComputeEngine.getDictionary('arithmetic'));
engine.evalue(['Add', 5, 2]);
```

## Comparing

Use the `compare()` function to compare two expressions, or the shorthands
`equal()`, `less()`, `lessEqual()`, `greater()`, `greaterEqual()`.

The comparison between expressions is structural so that $$x + 1$$ is not equal
to $$1 + x$$. To obtain the desired result, you may need to apply a canonical
form to the expressions, or evaluate them.

```js
const engine = new ComputeEngine();

console.log(engine.equal(['Add', 'x', 1], ['Add', 'x', 1]));
// -> true: the two expressions are equal

console.log(engine.equal(['Add', 'x', 1], ['Add', 1, 'x']));
// -> false: the two expressions are **not** equal

console.log(engine.equal(
  engine.canonical(['Add', 'x', 1]),
  engine.canonical(['Add', 1, 'x'])
);
// -> true: the two expressions are equal

console.log(engine.equal(
  ['Add', 2, 2],
  ['Add', 3, 1]
);
// -> false: the two expressions are **not** equal

console.log(engine.equal(
  engine.evaluate(['Add', 2, 2]),
  engine.evaluate(['Add', 3, 1])
);
// -> true: the two expressions are equal
```

## Domains

To obtain the domain of an expression, use the `domain()` function.

```js
const engine = new ComputeEngine();

engine.domain('Pi');
// -> "TranscendentalNumber"

engine.domain(['Add', 5, 2]);
// -> "Number"

engine.domain(engine.evaluate(['Add', 5, 2]));
// -> "PrimeNumber"
```
