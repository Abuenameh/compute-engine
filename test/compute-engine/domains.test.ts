import { ComputeEngine } from '../../src/compute-engine';
import type { Expression } from '../../src/public';

export const engine = new ComputeEngine();

/*
    [a, b, compare(a,b)]
*/

const domains: Expression[] = [
  0,
  5,
  'Integer',
  'RationalNumber',
  'AlgebraicNumber',
  'RealNumber',
  'ComplexNumber',
  'Number',
  ['Range', 0, 0],
  ['Range', 1, 1],
  ['Range', -Infinity, Infinity],
  ['Range', 1, 5],
  ['Interval', 0, 0],
  ['Interval', -Infinity, Infinity],
  ['Interval', 0, Infinity],
  ['Interval', 0, 5],
  ['And', ['Greater', 0], ['Lesser', 5]],
  ['Union', ['Range', 1, 5], ['Range', 5, Infinity]],
];

describe('canonicalDomainForm', () => {
  // https://jestjs.io/docs/next/api#testeachtablename-fn-timeout
  test.each(domains)('canonicalDomainForm("%p")', (domain) => {
    expect(engine.format(domain, 'canonical-domain')).toMatchSnapshot();
  });
});
