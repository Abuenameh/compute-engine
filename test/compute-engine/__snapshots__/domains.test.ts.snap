// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`canonicalDomainForm canonicalDomainForm(""AlgebraicNumber"") 1`] = `"AlgebraicNumber"`;

exports[`canonicalDomainForm canonicalDomainForm(""ComplexNumber"") 1`] = `"ComplexNumber"`;

exports[`canonicalDomainForm canonicalDomainForm(""Integer"") 1`] = `"Integer"`;

exports[`canonicalDomainForm canonicalDomainForm(""Number"") 1`] = `"Number"`;

exports[`canonicalDomainForm canonicalDomainForm(""RationalNumber"") 1`] = `"RationalNumber"`;

exports[`canonicalDomainForm canonicalDomainForm(""RealNumber"") 1`] = `"RealNumber"`;

exports[`canonicalDomainForm canonicalDomainForm("["And", [Array], [Array]]") 1`] = `
Array [
  "And",
  Array [
    "Greater",
    0,
  ],
  Array [
    "Lesser",
    5,
  ],
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Interval", -Infinity, Infinity]") 1`] = `
Array [
  "Interval",
  -Infinity,
  Infinity,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Interval", 0, 0]") 1`] = `
Array [
  "Interval",
  0,
  0,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Interval", 0, 5]") 1`] = `
Array [
  "Interval",
  0,
  5,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Interval", 0, Infinity]") 1`] = `
Array [
  "Interval",
  0,
  Infinity,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Range", -Infinity, Infinity]") 1`] = `
Array [
  "Range",
  -Infinity,
  Infinity,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Range", 0, 0]") 1`] = `
Array [
  "Range",
  0,
  0,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Range", 1, 1]") 1`] = `
Array [
  "Range",
  1,
  1,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Range", 1, 5]") 1`] = `
Array [
  "Range",
  1,
  5,
]
`;

exports[`canonicalDomainForm canonicalDomainForm("["Union", [Array], [Array]]") 1`] = `
Array [
  "Union",
  Array [
    "Range",
    1,
    5,
  ],
  Array [
    "Range",
    5,
    Infinity,
  ],
]
`;

exports[`canonicalDomainForm canonicalDomainForm("0") 1`] = `0`;

exports[`canonicalDomainForm canonicalDomainForm("5") 1`] = `5`;
